diff -urN rdesktop-1.0.0.org/rdesktop.h rdesktop-1.0.0/rdesktop.h
--- rdesktop-1.0.0.org/rdesktop.h	Tue May  8 22:40:12 2001
+++ rdesktop-1.0.0/rdesktop.h	Wed May  9 11:04:49 2001
@@ -31,6 +31,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <endian.g>
 
 #define VERSION "1.0.0-pl19-6"
 
@@ -38,22 +39,19 @@
 /*
  * check endianess
  */
-
-#if defined(__i386__) || defined(__alpha__) || defined(__vax__) || \
-      defined(_AIX)
-#define L_ENDIAN
-#elif defined(__sparc) || defined(__hppa) || defined(__mips) || \
-      defined(__PPC__) || defined(__m68k__)
-#define B_ENDIAN
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+# define L_ENDIAN
+#elif __BYTE_ORDER == __BIG_ENDIAN
+# define B_ENDIAN
 #else
 #error Unknown endianness. Edit rdesktop.h.
-#endif
+#endif                                    
 
 /*
  * check if we need to align data
  */
 
-#if defined(__sparc) || defined(__alpha__) || defined(__hppa) || \
+#if defined(__sparc__) || defined(__alpha__) || defined(__hppa__) || \
     defined(_AIX)
 #define NEED_ALIGN
 #endif
